_posts/2019-07-24-pattern-matching-expert-system-classification.md

```diff
@@ -0,0 +1,92 @@
+
+- food > fruit > apple > gala
+- fruit and vegetables > fruit > apple
+- food > produce > fresh fruit > apples
+
+In our case all of these must be mapped to the target category "Fruit > Apples". How could we describe an expert rule that tells the system to map products with any of the listed paths to their target category?
+
+> If any part of the category contains "fruit" and a following part is either "apple" or "apples" then the target category is "Fruit > Apples".
+
+Imagine a massive registry of rules in this form. It will be hard to find specific rules and writing them all out like this can be time consuming. Not to mention that in practice the rules are rarely as simple as in this example. We need a better manageable way to define rules.
+
+## Pattern matching
+
+Pattern matching is a technique used to match a pattern against sequences of tokens. The pattern describes what a sequence must look like in order to match. Probably the most common example of pattern matching is _regular expressions_. Using a regular expression the conditions from the above rule can be expressed as:
+
+```
+fruit.*>\s*apples?(\s*>|$)
+```
+
+There are some problems with regular expressions that make it a suboptimal choice for describing our rules. They are hard to learn, they quickly become complex and hard to read, and they are too generic and powerful. All of this increases the chance of mistakes.
+
+An interesting alternative is PostgreSQL's [ltree](https://www.postgresql.org/docs/current/ltree.html) extension. It features a language specifically designed to match tree paths. Unfortunately it does not have all the features that we require. And it has been implemented for PostgreSQL only. But if that's what you're using it might be worth taking a look at.
+
+### Tiered Category Expressions
+
+We ended up creating a custom pattern matching language, similar to ltree's, but tailored to our needs. We have named it _Tiered Category Expressions_ (TCEs).
+
+It is designed to be easy to learn for non-technical users, and easy to read and understand. In the simplest scenario an expression mimics the category it is supposed to match, e.g. "fruit > apple" can be matched with the expression `fruit > apple`.
+
+Using a TCE the above rule is expressed as:
+
+```
+>> %fruit% >> apple | apples
+```
+
+TCE features a.o. child and descendant matching, wildcard matching, and negation. Expressions match subtrees by default. Whitespace and special characters are ignored and matching is always case insensitive. Read more about these and other language features in the [TCE language reference](https://github.com/q-m/tiered_category_expressions/blob/master/LANGREF.md). Also TCEs can be transpiled to regular expressions, so the actual matching can be delegated to any platform that supports regular expressions. The Ruby implementation can be found [here](https://github.com/q-m/tiered_category_expressions).
```

Comment:
Is it clear what a _subtree_ is?

Reply:
Maybe not entirely, but it's not important. Readers may navigate to the lang ref for more details and examples.